pydoit.md

```diff
@@ -0,0 +1,138 @@
+-----------------------------
+
+Complex workflows may require several machines run the actions. Reasons include:
+- Several scientists want to work on the same workflow
+- Compute tools or postprocessing tools are not available
+    - licensing
+    - *just reviewing* the text of a document on a laptop
+    - HPC machine
+
+`pydoit` builds a database (in the folder of the `doit.py`) that tracks (via hashes of the *dependencies*) which *actions* need to be rerun. That means that each task has to be executed at least once, even if the *targets* already exist or are up-to-date.
+Problems arise, if due to some of the reasons above, the current machine is unable to do so. 
+
+One remedy is to _store_ a *target* in our version control system and tell `doit`: 
+*Trust me, the target in the version control is up-to-date.*
+
+### Proposed setup:
+~~~py
+    /data/dodo.py           # generates input data
+    /simulations/dodo.py    # turns input data into results
+    /images/dodo.py         # turns results into images
+    /paper/dodo.py          # turns images into a paper
+    /dodo.py       # optionally connect all tasks
+~~~
+
+The idea is to decouple the whole workflow into multiple `dodo` files. Someone, who just wants to build the `paper.pdf`, relies on up-to-date files in `/images/` and can simply run `doit` in the `/paper/` directory. 
+If one wants to execute the _full_ workflow, calling the root `doit` in `/`, will trigger the complete workflow.
+
+If you work with _absolute_ paths in your `dodo` files, such as ...
+~~~py
+# /simulation/dodo.py
+from pathlib import Path
+def task_simulation_1():
+    simulation_dir = Path(__file__).absolute().parent
+    script = str(simulation_dir / "simulation_1.py")
+    input = str(simulation_dir.parent / "data" / "input_1.dat")
+    image = str(simulation_dir.parent / "images" / "graph_1.pdf")
+    return {
+        "file_dep": [script, input],
+        "targets": [image],
+        "actions": [f"{python3 {script} {input}]
+    }
+~~~
+
+... and ...
+~~~py
+# /paper/dodo.py
+from pathlib import Path
+def task_simulation_1():
+    paper_dir = Path(__file__).absolute().parent
+    tex = str(paper_dir / "paper.tex")
+    pdf = str(paper_dir / "paper.pdf")
+    image = str(paper_dir.parent / "images" / "graph_1.pdf")
+    return {
+        "file_dep": [tex, image],
+        "targets": [pdf],
+        "actions": [f"{pdflatex {tex}]
+    }
+~~~
+... that makes it easy to connect the `image` target of the simulation task with the same `image` dependency of the paper task by just importing both into the same _root_ file
+
+~~~py
+# (root) /dodo.py
+from simulation.dodo import *
+from paper.dodo import *
+~~~
+
+
+Skipping huge simulations
+-------------------------
+
+If you just want to skip tasks (e.g. if they are too time consuming) you can define a flag and skip the task definition, if that flag is set. The simplest way would be to manually set the flag in the `dodo.py` file like
```

Comment:
You can also use the `ignore` command of doit: `doit ignore huge_simulation` and it will just skip it until you say `doit forget huge_simulation`.

Reply:
I played around with this option, but it will break the chain. So every task that has the target of "huge_simulation" as dependency will not be built. 
A -> B -> C -> D.
if you `doit ignore B`, it will also not run C and D. 

Or maybe I misunderstood?

Reply:
Maybe I misunderstood. You have already "cut up" your dependency graph with the subfolders. Why do you need another option to cut it up?

Reply:
First, these are all just recommendations or options you could use. 

Use case 1: You want to write the paper without the simulation software installed --> subfolder

Use case 2: You want to modify your inputs and conveniently update the figures and the paper but simulation_7 takes 3 hours --> skip